plugins/http_plugin/beast_http_session.hpp

```diff
@@ -0,0 +1,587 @@
+// see https://stackoverflow.com/questions/58453017/boost-asio-tcp-socket-1-70-not-backward-compatible
+#if BOOST_VERSION < 107000                
+    typedef tcp::socket tcp_socket_t;
+#else
+    typedef asio::basic_stream_socket<asio::ip::tcp, asio::io_context::executor_type> tcp_socket_t;
+#endif
+
+    using boost::asio::local::stream_protocol;
+    
+#if BOOST_VERSION < 107000
+    using local_stream = boost::asio::basic_stream_socket<stream_protocol>;
+#else 
+#if BOOST_VERSION < 107300
+    using local_stream = beast::basic_stream<
+    stream_protocol,
+    asio::executor,
+    beast::unlimited_rate_policy>;
+#else      
+using local_stream = beast::basic_stream<
+    stream_protocol,
+    asio::any_io_executor,
+    beast::unlimited_rate_policy>;    
+#endif    
+#endif
+
+    //------------------------------------------------------------------------------
+    // fail() 
+    //  this function is generally reserved in the case of a severe error which results 
+    //  in immediate termiantion of the session, with no response sent back to the client
+    //  currently also includes SSL "short read" error for security reasons:
+    // 
+    //  https://github.com/boostorg/beast/issues/38
+    //  https://security.stackexchange.com/questions/91435/how-to-handle-a-malicious-ssl-tls-shutdown
+    void fail(beast::error_code ec, char const* what)
+    {
+        fc_elog(logger, "${w}: ${m}", ("w", what)("m", ec.message()));
+        fc_elog(logger, "closing connection");
+    }
+
+    // this needs to be declared outside of the beast_http_session, because it doesn't work for 
+    // UNIX socket session, since it lacks address() function in endpoint
+    //  Since this is a virtual function, std::enable_if_t/SFNIAE doesn't work 
+    template<class T>
+    bool allow_host(const http::request<http::string_body>& req, T& session, std::shared_ptr<http_plugin_state> plugin_state) {
+        auto is_conn_secure = session.is_secure();
+
+        auto& socket = session.socket();
+#if BOOST_VERSION < 107000                
+        auto& lowest_layer = beast::get_lowest_layer<tcp_socket_t&>(socket);
+#else
+        auto& lowest_layer = beast::get_lowest_layer(socket);
+#endif                
+        auto local_endpoint = lowest_layer.local_endpoint();
+        auto local_socket_host_port = local_endpoint.address().to_string() 
+                + ":" + std::to_string(local_endpoint.port());
+        const auto& host_str = req["Host"].to_string();
+        if (host_str.empty() 
+            || !host_is_valid(*plugin_state, 
+                                host_str, 
+                                local_socket_host_port, 
+                                is_conn_secure)) 
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    // Handle HTTP conneciton using boost::beast for TCP communication
+    // This uses the Curiously Recurring Template Pattern so that
```

Comment:
This is not an example of the `Curiously Recurring Template Pattern`. Ah, you mean derived classes can use the CRTP with this class.